engine/src/main/java/com/datatorrent/stram/util/LoggerUtil.java

```diff
@@ -464,4 +474,55 @@ public static boolean removeAppender(String name)
     }
     return names;
   }
+
+  /**
+   * Makes MDC properties
+   */
+  public static void setupMDC()
+  {
+    String node = StramClientUtils.getHostName();
+    MDC.put("node", node == null ? "unknown" : node);
+
+    String value = System.getenv("HADOOP_USER_NAME");
+    if (value != null) {
+      MDC.put("user", value);
+    }
+
+    value = System.getenv("CONTAINER_ID");
+    if (value != null) {
+      MDC.put("containerId", value);
+    }
+
+    value = System.getProperties().getProperty(APPLICATION_NAME.getLongName());
+    if (value != null) {
+      MDC.put("application", value);
+    }
+
+    if (MDC.get("service") == null) {
+      MDC.put("service", getService());
+    }
+  }
+
+  /**
+   * Returns the service name
+   * @return Service name
+   */
+  private static String getService()
```

Comment:
Unfortunately it is impossible to prevent all possible stupid mistakes. We can only try to minimize the number of mistakes. And if we are going to make it the setting of MDC properties MUST be done in the thread with the main method in the root of the program stack only because the thread of the main class is a parent of all other threads.

I've added the check that the implicit MDC initialization is called from the main thread.

    StackTraceElement[] stack = Thread.currentThread().getStackTrace();
    if (!"main".equals(stack[stack.length - 1].getMethodName())) {
      throw new RuntimeException("The implicit initialization of MDC properties must be done from the main thread only");
    }

But if another developer modifies "main" class for the master or workers and introduces another "main" class, it will be easy checked and fixed, because the service name in logs will have the value "unknown".